packages/components/src/internal/url/AppURL.ts

```diff
@@ -171,61 +135,112 @@ export class AppURL {
         });
     }
 
-    addParam(key: string, value: URLParam, includeEmptyParams?: boolean): AppURL {
-        return this.addParams({ [key]: value }, includeEmptyParams);
+    addParam(key: string, value: QueryParamValue): AppURL {
+        return this.addParams({ [key]: value });
     }
 
-    addParams(params: Record<string, URLParam>, includeEmptyParams = false): AppURL {
+    addParams(params: Record<string, QueryParamValue>): AppURL {
         if (params) {
-            const encodedParams: Record<string, string> = {};
-            Object.keys(params).forEach(key => {
-                const value = params[key];
-                if (includeEmptyParams || (value !== undefined && value !== null)) {
-                    encodedParams[encodeURIComponent(key)] = encodeURIComponent(value);
-                }
-            });
-            return new AppURL({
-                ...this,
-                _params: this._params ? { ...this._params, ...encodedParams } : encodedParams,
-            });
+            const nonEmptyParams = Object.keys(params).reduce(
+                (result, key) => {
+                    const value = params[key];
+                    if (value !== null && value !== undefined) result[key] = value;
+                    return result;
+                },
+                { ...this._params } as Record<string, QueryParamValue>
+            );
+            return new AppURL({ ...this, _params: nonEmptyParams });
         }
 
         return this;
     }
 
-    toHref(urlPrefix?: string): string {
-        return '#' + this.toString(urlPrefix);
+    setContainerPath(containerPath: string): AppURL {
+        return new AppURL({ ...this, _containerPath: containerPath });
     }
 
-    toQuery(urlPrefix?: string): Record<string, string> {
-        const query = { ...this._params };
+    getContainerPath(): string {
+        if (this._containerPath === undefined) return ActionURL.getContainer();
+        return this._containerPath;
+    }
 
-        if (this._filters) {
-            this._filters.forEach(f => {
-                query[f.getURLParameterName(urlPrefix)] = f.getURLParameterValue();
-            });
-        }
+    setProductId(productId: string): AppURL {
+        return new AppURL({ ...this, _productId: productId });
+    }
 
-        return query;
+    getProductId(): string {
+        if (this._productId === undefined) return getPrimaryAppProductId() ?? '';
+        return this._productId;
     }
 
-    toString(urlPrefix?: string): string {
-        const url = this._baseUrl;
-        const parts = [];
+    /**
+     * Returns true if the AppURL points to a path for the current app in the current container
+     */
+    isAppPath(): boolean {
+        const currentContainerPath = ActionURL.getContainer();
```

Comment:
(Tangential) It's a bit unfortunate how `ActionURL` is set up to compute these "properties" of the URL. It ends up needing to run through `getPathFromLocation()` for each call making these "getters" not cheap (compared a normal property getter).